fix_discount_closure.py

```
@@ -0,0 +1,40 @@
+    # If no header-level discount, check for item-level discounts
+    if not has_any_discount:
+        for item in self.items:
+            if item.discount_percentage > 0 or item.discount_amount > 0:
+                has_any_discount = True
+                break
+
+    if has_any_discount:
+        # Get unique sales orders linked to this invoice
+        linked_sales_orders = {item.sales_order for item in self.items if item.sales_order}
+        for so_name in linked_sales_orders:
+            so = frappe.get_doc("Sales Order", so_name)
```

Ruff (0.12.2)

Undefined name `frappe`

(F821)